File: MLR-spring/src/main/java/com/mlr/controller/AnalyticsController.java
package com.mlr.controller;

import com.mlr.service.AnalyticsService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/analytics")
@PreAuthorize("hasRole('USER')")
public class AnalyticsController {

    @Autowired
    private AnalyticsService analyticsService;

    @GetMapping("/spending")
    public ResponseEntity<?> getSpendingAnalytics(Authentication authentication) {
        String userId = authentication.getName();
        return ResponseEntity.ok(analyticsService.getSpendingAnalytics(userId));
    }

    @GetMapping("/leakage-alerts")
    public ResponseEntity<?> getLeakageAlerts(Authentication authentication) {
        String userId = authentication.getName();
        return ResponseEntity.ok(analyticsService.getLeakageAlerts(userId));
    }

    @PostMapping("/leakage-alerts/{id}/read")
    public ResponseEntity<?> markAlertAsRead(@PathVariable Long id, Authentication authentication) {
        String userId = authentication.getName();
        boolean result = analyticsService.markAlertAsRead(userId, id);

        if (!result) {
            return ResponseEntity.status(404).body(Map.of("message", "Alert not found or access denied"));
        }
        return ResponseEntity.ok(Map.of("success", true, "message", "Alert marked as read"));
    }

    @GetMapping("/leakage")
    public ResponseEntity<?> getLeakageAnalytics(Authentication authentication) {
        String userId = authentication.getName();
        return ResponseEntity.ok(analyticsService.getLeakageAnalytics(userId));
    }
}

File: MLR/AnalyticsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using MLR.Services;

namespace MLR.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize(Roles = "USER")]
public class AnalyticsController : ControllerBase
{
    private readonly IAnalyticsService _analyticsService;

    public AnalyticsController(IAnalyticsService analyticsService)
    {
        _analyticsService = analyticsService;
    }

    [HttpGet("spending")]
    public async Task<IActionResult> GetSpendingAnalytics()
    {
        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var analytics = await _analyticsService.GetSpendingAnalyticsAsync(userId);
        return Ok(analytics);
    }

    [HttpGet("leakage-alerts")]
    public async Task<IActionResult> GetLeakageAlerts()
    {
        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var alerts = await _analyticsService.GetLeakageAlertsAsync(userId);
        return Ok(alerts);
    }

    // New Endpoint to Mark Alert as Read
    [HttpPost("leakage-alerts/{id}/read")]
    public async Task<IActionResult> MarkAlertAsRead(int id)
    {
        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var result = await _analyticsService.MarkAlertAsReadAsync(userId, id);

        if (!result)
        {
            return NotFound(new { message = "Alert not found or access denied" });
        }

        return Ok(new { success = true, message = "Alert marked as read" });
    }

    [HttpGet("leakage")]
    public async Task<IActionResult> GetLeakageAnalytics()
    {
        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var analytics = await _analyticsService.GetLeakageAnalyticsAsync(userId);
        return Ok(analytics);
    }
}